sklearn/calibration.py

```diff
@@ -33,25 +33,31 @@
 from .utils.validation import _deprecate_positional_args
 
 
-def _fit_calibrated_classifer(estimator, X, y, train, test, supports_sw,
-                              method, classes, sample_weight=None):
+def _fit_calibrated_classifer(estimator, X, y, train, test,
+                              method, classes, sample_weight=None,
+                              base_estimator_uses_sw=False, **fit_params):
     """Fit calibrated classifier for a given dataset split.
 
     Returns
     -------
     calibrated_classifier : estimator object
         The calibrated estimator.
     """
-    if sample_weight is not None and supports_sw:
-        estimator.fit(X[train], y[train],
-                      sample_weight=sample_weight[train])
+    fit_params_train = {}
+    for key, val in fit_params.items():
+        fit_params_train[key] = _safe_indexing(val, train)
+
+    if base_estimator_uses_sw:
```

Comment:
```python
X_train, y_train = _safe_indexing(X, train), _safe_indexing(y, train)
X_test, y_test = _safe_indexing(X, test), _safe_indexing(y, test)
if sample_weight is not None:
    sample_weight_train = _safe_indexing(sample_weight, train)
    sample_weight_test = _safe_indexing(sample_weight, test)
    fit_parameters = signature(base_estimator.fit).parameters
    if "sample_weight" in fit_parameters:
        estimator.fit(
            X_train, y_train, sample_weight=sample_weight_train,
            **fit_params_train,
        )
    else:
        estimator.fit(X_train, y_train, **fit_params_train)
calibrated_classifier = _CalibratedClassifier(
    estimator, method=method, classes=classes,
)
calibrated_classifier.fit(
    X_test, y_test, sample_weight=sample_weight_test
)
```

Reply:
The snippet doesn't quite work:

1. if `sample_weight` _is_ None, `estimator` is never fitted
2. on the last line, you use `sample_weight_test`, which is undefined if `sample_weight` is None

I tried to refactor according to the "spirit" of the snippet.